middle_end/flambda2/types/reify.ml

```diff
@@ -575,57 +577,99 @@ let reify ~allowed_if_free_vars_defined_in ~var_is_defined_at_toplevel
                   { contents = Known (Immutable { fields });
                     length = _;
                     alloc_mode;
-                    element_kind
+                    element_kinds
                   })
           }) -> (
       match fields with
       | [||] -> (
-        match element_kind with
-        | Ok element_kind ->
+        match element_kinds with
+        | Ok element_kinds ->
           let array_kind =
-            Empty_array_kind.of_element_kind
-              (Flambda_kind.With_subkind.kind element_kind)
+            Empty_array_kind.of_element_kinds
+              (List.map Flambda_kind.With_subkind.kind element_kinds)
           in
           Lift (Empty_array array_kind)
         | Unknown | Bottom -> try_canonical_simple ())
       | _ -> (
         let fields = Array.to_list fields in
-        match element_kind with
+        match element_kinds with
         | Unknown -> try_canonical_simple ()
         | Bottom ->
           (* CR someday vlaviron: we could use [Lift Empty_array] here *)
           try_canonical_simple ()
-        | Ok element_kind -> (
-          let kind = Flambda_kind.With_subkind.kind element_kind in
-          match kind with
-          | Value -> (
-            let field_types_and_expected_kinds =
-              List.map (fun ty -> ty, Flambda_kind.value) fields
-            in
+        | Ok element_kinds -> (
+          let kinds = List.map Flambda_kind.With_subkind.kind element_kinds in
+          match kinds with
+          | [] -> Misc.fatal_error "Did not expect empty element kinds"
+          | _ :: _ :: _ -> (
+            let field_types_and_expected_kinds = List.combine fields kinds in
             match
               try_to_reify_fields env ~var_allowed alloc_mode
                 ~field_types_and_expected_kinds
             with
-            | Some fields -> Lift (Immutable_value_array { fields })
+            (* XXX mshinwell: We need to convince ourselves that there is no
+               problem with making different decisions based on immediacy
```

Comment:
I assume this means static arrays of immediates will become non-scannable blocks? I can't think of any reason that'd be a problem, but may be worth considering stuff like [this](https://github.com/ocaml-multicore/saturn/blob/306bea620cc0cfcc33639c45a56da59add9bdd92/src/ws_deque.ml#L45)